porch/pkg/git/package.go

```diff
@@ -152,36 +157,27 @@ func (p *gitPackageRevision) GetResources(ctx context.Context) (*v1alpha1.Packag
 }
 
 func (p *gitPackageRevision) GetUpstreamLock() (kptfile.Upstream, kptfile.UpstreamLock, error) {
-	repo, err := p.parent.getRepo()
+	resources, err := p.GetResources(context.Background())
 	if err != nil {
-		return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot determine package lock: %w", err)
+		return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot determine package lock; cannot retrieve resources: %w", err)
 	}
 
-	if p.ref == nil {
-		return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot determine package lock; package has no ref")
+	// get the upstream package URL from the Kptfile
+	var kf *kptfile.KptFile
+	for k, v := range resources.Spec.Resources {
+		if k == kptfile.KptFileName {
+			var err error
+			kf, err = pkg.DecodeKptfile(strings.NewReader(v))
+			if err != nil {
+				return kptfile.Upstream{}, kptfile.UpstreamLock{}, fmt.Errorf("cannot read package lock from Kptfile: %w", err)
```

Comment:
This error message maybe seems a bit misleading, but the nested error might explain to the user that the format of the Kptfile is somehow invalid so it can't be decoded.

Reply:
Makes sense to me! I changed the message to `cannot decode Kptfile`